File: Front/src/api/endpoints/sales.ts
// 판매 데이터 업로드 API — api_spec.md §6.
import { api } from "../../lib/api";

export interface CSVUploadResponse {
  imported: number;
  skipped: number;
  skipped_reasons: string[];
  anomaly_count: number;
  auto_created_menus: number;
}

export interface CSVUploadColumns {
  date_column: string;
  menu_column: string;
  quantity_column: string;
  price_column: string;
  external_sale_id_column?: string | null;
}

export interface CSVUploadOptions {
  /** true 시 매장 메뉴에 없는 항목을 카테고리='자동등록'으로 즉시 추가. */
  auto_create_menus?: boolean;
}

export async function uploadSalesCsv(
  file: File,
  columns: CSVUploadColumns,
  options: CSVUploadOptions = {},
): Promise<CSVUploadResponse> {
  const form = new FormData();
  form.append("file", file);
  form.append("date_column", columns.date_column);
  form.append("menu_column", columns.menu_column);
  form.append("quantity_column", columns.quantity_column);
  form.append("price_column", columns.price_column);
  if (columns.external_sale_id_column) {
    form.append("external_sale_id_column", columns.external_sale_id_column);
  }
  if (options.auto_create_menus) {
    form.append("auto_create_menus", "true");
  }
  // 업로드는 시간이 걸릴 수 있어 ky 기본 타임아웃(10s)을 늘림.
  return api.post("sales/upload", { body: form, timeout: 120_000 }).json<CSVUploadResponse>();
}

export interface SalesSummary {
  total_revenue: number;
  total_sales_count: number;
  this_month_revenue: number;
  this_month_sales_count: number;
  today_revenue: number;
  today_sales_count: number;
  last_sale_at: string | null;
}

export async function getSalesSummary(): Promise<SalesSummary> {
  return api.get("sales/summary").json<SalesSummary>();
}

export interface MonthlyRevenuePoint {
  year_month: string;
  revenue: number;
  sales_count: number;
}

export async function getMonthlyRevenue(months = 6): Promise<MonthlyRevenuePoint[]> {
  return api.get("sales/monthly", { searchParams: { months } }).json<MonthlyRevenuePoint[]>();
}

export interface TopMenuItem {
  menu_name: string;
  quantity: number;
  revenue: number;
}

export async function getTopMenus(limit = 5): Promise<TopMenuItem[]> {
  return api.get("sales/top-menus", { searchParams: { limit } }).json<TopMenuItem[]>();
}

export interface DailyRevenuePoint {
  date: string;
  revenue: number;
  sales_count: number;
}

export async function getDailyRevenue(days = 7): Promise<DailyRevenuePoint[]> {
  return api.get("sales/daily", { searchParams: { days } }).json<DailyRevenuePoint[]>();
}

export async function getDailyRevenueForMonth(yearMonth: string): Promise<DailyRevenuePoint[]> {
  return api
    .get("sales/daily-by-month", { searchParams: { year_month: yearMonth } })
    .json<DailyRevenuePoint[]>();
}

export interface WeeklyRevenuePoint {
  week_label: string;
  revenue: number;
  sales_count: number;
}

export async function getWeeklyRevenueThisMonth(): Promise<WeeklyRevenuePoint[]> {
  return api.get("sales/weekly-this-month").json<WeeklyRevenuePoint[]>();
}

File: Back/app/schemas/sales.py
"""판매 API 스키마 — api_spec §6."""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class CSVUploadResponse(BaseModel):
    imported: int
    skipped: int
    skipped_reasons: list[str] = Field(default_factory=list)
    anomaly_count: int = 0
    auto_created_menus: int = 0


class SalesSummaryResponse(BaseModel):
    """홈 화면 매출 요약 카드용 — 대시보드 조회 API가 별도로 없어 최소 집계만 제공."""

    total_revenue: int
    total_sales_count: int
    this_month_revenue: int
    this_month_sales_count: int
    today_revenue: int
    today_sales_count: int
    last_sale_at: datetime | None


class MonthlyRevenuePoint(BaseModel):
    year_month: str  # "2026-05"
    revenue: int
    sales_count: int


class DailyRevenuePoint(BaseModel):
    date: str  # "2026-07-28"
    revenue: int
    sales_count: int


class WeeklyRevenuePoint(BaseModel):
    week_label: str  # "1주차"
    revenue: int
    sales_count: int


class TopMenuItem(BaseModel):
    menu_name: str
    quantity: int
    revenue: int
